README.md
lua-snowflake
=============

An implementation of [Snowflake](https://blog.twitter.com/2010/announcing-snowflake) for Lua. Snowflake is an algorithm
which supports ordered, distributed id generation. 

Installation
------------

    $ luarocks install snowflake
     
Usage
-----

    local sf = require "snowflake"
    sf.init(0x01, 0x01)
    local uuid = sf.next_id()

`uuid` will be a 64-bit number represented as a string that is structured as follows:

    6  6         5         4         3         2         1         
    3210987654321098765432109876543210987654321098765432109876543210
    
    tttttttttttttttttttttttttttttttttttttttttdddddnnnnnsssssssssssss

where default

* `s` is a 12-bit integer that increments when next_id() is called multiple times for the same millisecond
* `n` is a 5-bit integer representing the node within a given data center
* `d` is a 5-bit integer representing a unique data center or group of servers
* `t` is a 42-bit integer representing the current timestamp in milliseconds
    * the number of milliseconds to have elapsed since 1413817200000 or 2014-10-20T15:00:00.000Z

`sf.init(datacenter_id, node_id, snowflake_epoc, node_id_bits, datacenter_id_bits, sequence_bits, delta_offset)` is used to initialize snowflake and set values for ddddd and nnnnn as follows:

* `datacenter_id` is an integer `d`, where 0 ≤ d < (1 << datacenter_id_bits) and specifies the `ddddd` portion of the id.
* `node_id` is an integer `n`, where 0 ≤ n < (1 << node_id_bits) and specifies the `nnnnn` portion of the id.
* `snowflake_epoc` the start timestamp is expressed in milliseconds, the part beyond delta_offset will be ignored.
* `node_id_bits` is an integer representing the number of bits occupied by a node in a given data center.
* `datacenter_id_bits` is an integer representing the number of bits occupied by a unique data center or a group of servers
* `sequence_bits` is an integer, the number of bits occupied by the number of calls in the same millisecond.
* `delta_offset` timestamp delta offset (unit: milliseconds) [`delta_offset=1` 1 millisecond, `delta_offset=10` 10 milliseconds, `delta_offset=100` 100 milliseconds, `delta_offset=1000` 1 second`]


CHANGELOG.md
# v2.1-1

- feat: support timestamp delta offset configuration

test_sonyflake.lua
-- 
-- The reason for "luaL_checklong" does not guarantee that every test is correct
--

local sf = require "snowflake"
local socket = require("socket")

function sleep(n)
    socket.select(nil, nil, n)
 end

sf.init(0x0, 0x1, 1609459200000, 8, 8, 8, 10)
local function check (sf ,start)
    local a = 1000
    local i = 0
    local last_offset
    while( a>0 )
    do
        local offset = sf.getmillisecond() - start
        local id = sf.next_id()
        if last_offset ~= offset then
            i = 0
        end
        local result =  sf.bit_split(id)
        if ("1, "..offset..", " .. i ~= result) then
            print("result:" .. result)
            print("offset:" .. offset)
            print("\n1, "..offset..", " .. i)
        end
        sleep(0.02)
        last_offset = offset
        a = a-1
        i = i+1
    end
end
check(sf, 160945920000)
print("success\n")

test_UidGenerator.lua
local sf = require "snowflake"
local socket = require("socket")

function sleep(n)
    socket.select(nil, nil, n)
 end

 local function check (sf ,start)
    local a = 1000
    local i = 0
    local last_offset
    while( a>0 )
    do
        local offset = sf.getmillisecond() - start
        local id = sf.next_id()
        if last_offset ~= offset then
            i = 0
        end
        local result =  sf.bit_split(id)
        if ("1, "..offset..", " .. i ~= result) then
            print("result:" .. result)
            print("offset:" .. offset)
            print("\n1, "..offset..", " .. i)
        end
        sleep(0.02)
        last_offset = offset
        a = a-1
        i = i+1
    end
end

sf.init(0x0, 0x1, 1609459200000, 11, 11, 13, 1000)
check(sf, 1609459200)
print("success\n")

src/main.c
#include <lua.h>
#include <lauxlib.h>
#include <stdbool.h>
#include <sys/time.h>
#include <math.h>
#include <stdlib.h>

#if LUA_VERSION_NUM < 502
#define luaL_newlib(L, l) (lua_newtable(L), luaL_register(L, NULL, l))
#endif

static bool initialized = false;

static int g_datacenter_id = 0;
static int g_node_id = 0;
static long g_last_timestamp = -1;
static int g_sequence = 0;

// 2014-10-20T15:00:00.000Z
#define SNOWFLAKE_EPOC 1413817200000L

#define NODE_ID_BITS 5
#define DATACENTER_ID_BITS 5
#define SEQUENCE_BITS 12
#define DELTA_OFFSET 1

struct conf {
    long snowflake_epoc;
    int node_id_bits;
    int datacenter_id_bits;
    int sequence_bits;
    int node_id_shift;
    int datacenter_id_shift;
    int timestamp_shift;
    int sequence_mask;
    int delta_offset;
} conf;

static long get_timestamp() {
    struct timeval tv;
    gettimeofday(&tv, NULL);
    switch (conf.delta_offset)
    {
    case 1000:
        return tv.tv_sec;
    case 100:
        return tv.tv_sec * 10 + tv.tv_usec / 100000;
    case 10:
        return tv.tv_sec * 100 + tv.tv_usec / 10000;
    case 1:
        return tv.tv_sec * 1000 + tv.tv_usec / 1000;
    }
    return tv.tv_sec * 1000 + tv.tv_usec / 1000;
}

static long get_til_next_millis(long last_timestamp) {
    long ts = get_timestamp();
    while (ts < last_timestamp) {
        ts = get_timestamp();
    }

    return ts;
}

static int luasnowflake_init(lua_State *L) {
    conf.snowflake_epoc = luaL_optlong(L, 3, SNOWFLAKE_EPOC);
    struct timeval tv;
    gettimeofday(&tv, NULL);
    if (conf.snowflake_epoc < 0 || conf.snowflake_epoc > tv.tv_sec * 1000) {
        char buffer[32];
        snprintf(buffer, 32, "%ld", tv.tv_sec * 1000);
        return luaL_error(L, "snowflake_epoc must be an long n where 1 ≤ n ≤ %s", buffer);
    }

    conf.node_id_bits = luaL_optint(L, 4, NODE_ID_BITS);
    if (conf.node_id_bits < 1) {
        return luaL_error(L, "node_id_bits must be an integer n where ≥ 1");
    }

    conf.datacenter_id_bits = luaL_optint(L, 5, DATACENTER_ID_BITS);
    if (conf.datacenter_id_bits < 1) {
        return luaL_error(L, "datacenter_id_bits must be an integer n where ≥ 1");
    }

    conf.sequence_bits = luaL_optint(L, 6, SEQUENCE_BITS);
    if (conf.sequence_bits < 1) {
        return luaL_error(L, "sequence_bits must be an integer n where ≥ 1");
    }

    conf.delta_offset = luaL_optint(L, 7, DELTA_OFFSET);
    switch (conf.delta_offset)
    {
    case 1000:
    case 100:
    case 10:
    case 1:
        break;
    default:
        return luaL_error(L, "does not support delta_offset: %d", conf.delta_offset);
    }

    conf.snowflake_epoc = conf.snowflake_epoc / conf.delta_offset;

    g_datacenter_id = luaL_checkint(L, 1);
    int max_datacenter_id_bits = (1 << conf.datacenter_id_bits) - 1;
    if (g_datacenter_id < 0x00 || g_datacenter_id > max_datacenter_id_bits) {
        return luaL_error(L, "datacenter_id must be an integer n, where 0 ≤ n ≤ %d", max_datacenter_id_bits);
    }

    g_node_id = luaL_checkint(L, 2);
    int max_node_id_bits = (1 << conf.node_id_bits) -1;
    if (g_node_id < 0x00 || g_node_id > max_node_id_bits) {
        return luaL_error(L, "node_id must be an integer n where 0 ≤ n ≤ %d", max_node_id_bits) ;
    }

    conf.node_id_shift = conf.sequence_bits;
    conf.datacenter_id_shift = (conf.node_id_shift + conf.node_id_bits);
    conf.timestamp_shift = (conf.datacenter_id_shift + conf.datacenter_id_bits);
    conf.sequence_mask = (0xffffffff ^ (0xffffffff << conf.sequence_bits));

    initialized = true;

    return 1;
}

static int luasnowflake_next_id(lua_State *L) {
    if (!initialized) {
        return luaL_error(L, "snowflake.init must be called first");
    }

    long ts = get_timestamp();

    if (g_last_timestamp == ts) {
        g_sequence = (g_sequence + 1) & conf.sequence_mask;
        if (g_sequence == 0) {
            ts = get_til_next_millis(g_last_timestamp);
        }
    } else {
        g_sequence = 0;
    }

    g_last_timestamp = ts;
    ts = ((ts - conf.snowflake_epoc) << conf.timestamp_shift)
            | (g_datacenter_id << conf.datacenter_id_shift)
            | (g_node_id << conf.node_id_shift)
            | g_sequence;

    char buffer[32];
    snprintf(buffer, 32, "%ld", ts);
    lua_pushstring(L, buffer);

    return 1;
}

static int bit_split(lua_State *L) {
    const char* snow_id_str = luaL_checkstring(L, 1);
    long int snow_id = atol(snow_id_str);
    long sequence = snow_id & ((long)1 << conf.sequence_bits) - 1;
    long data_machine_and = ((long)1 << (conf.node_id_bits + conf.datacenter_id_bits)) - 1;
    long data_machine_id = (snow_id >> conf.sequence_bits) & data_machine_and;
    int max = conf.sequence_bits + conf.node_id_bits + conf.datacenter_id_bits;
    long delta_offset_and = ((long)1 << (64 - max)) - 1;
    long delta_offset = (snow_id >> max) & delta_offset_and;
    char buffer[32];
    snprintf(buffer, 32, "%ld, %ld, %ld", data_machine_id, delta_offset, sequence);
    lua_pushstring(L, buffer);
    return 1;
}

static int getmillisecond(lua_State *L) {
    lua_pushnumber(L, get_timestamp());
    return 1;
}

static const struct luaL_Reg luasnowflake_lib[] = {
    {"init", luasnowflake_init},
    {"next_id", luasnowflake_next_id},
    {"getmillisecond", getmillisecond},
    {"bit_split", bit_split},
    {NULL, NULL},
};

LUALIB_API int luaopen_snowflake(lua_State *const L) {
    luaL_newlib(L, luasnowflake_lib);

    return 1;
}
